File: Python/006__ZigZag_Conversion.py
class Solution(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        if numRows == 1:
            return s

        direction = 1 # to control row# movement
        rowIndex = 0 # indicate which lane to add element
        result = ["" for _ in range(numRows)] # declare a nested list in which each row is a list

        for i in range(len(s)):
            result[rowIndex] += (s[i])
            rowIndex += direction
            if rowIndex == 0 or rowIndex == numRows - 1:
                direction *= -1
        return "".join(result)

File: Java/006__ZigZag_Conversion.java
class Solution {
    public String convert(String s, int numRows) {
        if (s == null || s.length() == 0) return "";
        if (numRows == 1) return s;
        HashMap<Integer, String> map = new HashMap<>();
        int idx = 0, dir = 1;
        for (char c : s.toCharArray()) {
            idx += dir;
            if (!map.containsKey(idx)) map.put(idx, new String());
            String temp = map.get(idx);
            temp += String.valueOf(c);
            map.put(idx, temp);
            if (idx == numRows) dir = -1;
            if (idx == 1) dir = 1;
        }
        String result = "";
        for (int i = 1; i <= numRows; i++) {
            if (map.get(i) != null) result += map.get(i);
        }
        return result;
    }
}
